File: test/ModularPipelines.UnitTests/Dependencies/OneWayDependenciesNonCollisionTests.cs
using ModularPipelines.Attributes;
using ModularPipelines.Context;
using ModularPipelines.Modules;
using ModularPipelines.TestHelpers;

namespace ModularPipelines.UnitTests.Dependencies;

public class OneWayDependenciesNonCollisionTests
{
    [Test]
    public async Task Modules_Not_Dependent_On_Each_Other_Succeed()
    {
        await Assert.That(() => TestPipelineHostBuilder.Create()
            .AddModule<DependencyConflictModule1>()
            .AddModule<DependencyConflictModule2>()
            .AddModule<DependencyConflictModule3>()
            .AddModule<DependencyConflictModule4>()
            .AddModule<DependencyConflictModule5>()
            .ExecutePipelineAsync()).ThrowsNothing();
    }

    [ModularPipelines.Attributes.DependsOn<DependencyConflictModule2>]
    private class DependencyConflictModule1 : Module<bool>
    {
        public override async Task<bool> ExecuteAsync(IModuleContext context, CancellationToken cancellationToken)
        {
            await Task.CompletedTask;
            return true;
        }
    }

    [ModularPipelines.Attributes.DependsOn<DependencyConflictModule3>]
    private class DependencyConflictModule2 : Module<bool>
    {
        public override async Task<bool> ExecuteAsync(IModuleContext context, CancellationToken cancellationToken)
        {
            await Task.CompletedTask;
            return true;
        }
    }

    [ModularPipelines.Attributes.DependsOn<DependencyConflictModule4>]
    private class DependencyConflictModule3 : Module<bool>
    {
        public override async Task<bool> ExecuteAsync(IModuleContext context, CancellationToken cancellationToken)
        {
            await Task.CompletedTask;
            return true;
        }
    }

    [ModularPipelines.Attributes.DependsOn<DependencyConflictModule5>]
    private class DependencyConflictModule4 : Module<bool>
    {
        public override async Task<bool> ExecuteAsync(IModuleContext context, CancellationToken cancellationToken)
        {
            await Task.CompletedTask;
            return true;
        }
    }

    private class DependencyConflictModule5 : Module<bool>
    {
        public override async Task<bool> ExecuteAsync(IModuleContext context, CancellationToken cancellationToken)
        {
            await Task.CompletedTask;
            return true;
        }
    }
}

File: test/ModularPipelines.UnitTests.FSharp/Dependencies/OneWayDependenciesNonCollisionTests.fs
namespace ModularPipelines.UnitTests.FSharp.Dependencies

open System.Threading.Tasks
open ModularPipelines.Attributes
open ModularPipelines.Context
open ModularPipelines.Enums
open ModularPipelines.Extensions
open ModularPipelines.Modules
open ModularPipelines.TestHelpers
open TUnit.Assertions
open TUnit.Assertions.Extensions
open TUnit.Assertions.FSharp.Operations
open TUnit.Core

module OneWayDependenciesNonCollisionTestModules =
    [<DependsOn(typeof<DependencyConflictModule2>)>]
    type DependencyConflictModule1() =
        inherit Module<bool>()
        override _.ExecuteAsync(_, _) = Task.FromResult(true)

    and [<DependsOn(typeof<DependencyConflictModule3>)>] DependencyConflictModule2() =
        inherit Module<bool>()
        override _.ExecuteAsync(_, _) = Task.FromResult(true)

    and [<DependsOn(typeof<DependencyConflictModule4>)>] DependencyConflictModule3() =
        inherit Module<bool>()
        override _.ExecuteAsync(_, _) = Task.FromResult(true)

    and [<DependsOn(typeof<DependencyConflictModule5>)>] DependencyConflictModule4() =
        inherit Module<bool>()
        override _.ExecuteAsync(_, _) = Task.FromResult(true)

    and DependencyConflictModule5() =
        inherit Module<bool>()
        override _.ExecuteAsync(_, _) = Task.FromResult(true)

type OneWayDependenciesNonCollisionTests() =
    [<Test>]
    member _.Modules_Not_Dependent_On_Each_Other_Succeed() = async {
        let! pipelineSummary =
            TestPipelineHostBuilder.Create()
                .AddModule<OneWayDependenciesNonCollisionTestModules.DependencyConflictModule1>()
                .AddModule<OneWayDependenciesNonCollisionTestModules.DependencyConflictModule2>()
                .AddModule<OneWayDependenciesNonCollisionTestModules.DependencyConflictModule3>()
                .AddModule<OneWayDependenciesNonCollisionTestModules.DependencyConflictModule4>()
                .AddModule<OneWayDependenciesNonCollisionTestModules.DependencyConflictModule5>()
                .ExecutePipelineAsync()
            |> Async.AwaitTask

        do! check(Assert.That(pipelineSummary.Status).IsEqualTo(Status.Successful))
    }
